main/src/library/Foldable.flix

```diff
@@ -0,0 +1,148 @@
+    ///
+    /// Returns `false` if `t` is empty.
+    ///
+    pub def exists[t :# Type -> Type : Foldable, a](f: a -> Bool, t: t[a]): Bool =
+        Foldable.foldLeft((acc, x) -> acc or f(x), false, t)
+
+    ///
+    /// Returns `true` if and only if all elements in `t` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `t` is empty.
+    ///
+    pub def forall[t :# Type -> Type : Foldable, a](f: a -> Bool, t: t[a]): Bool =
+        Foldable.foldLeft((acc, x) -> acc and f(x), true, t)
+
+    ///
+    /// Optionally returns the first element of `t` that satisfies the predicate `f` when searching from left to right.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def find[t :# Type -> Type : Foldable, a](f: a -> Bool, t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> match acc {
+            case None => if (f(x)) Some(x) else None
+            case Some(y) => Some(y)
+        }, None, t)
+
+    ///
+    /// Returns the result of mapping each element and combining the results.
+    ///
+    pub def foldMap[t :# Type -> Type : Foldable, a, b : Monoid](f: a -> b, t: t[a]): b =
+        Foldable.foldLeft((acc, x) -> Monoid.combine(f(x), acc), Monoid.empty(), t)
+
+    ///
+    /// Returns `t` with `a` inserted between every adjacent element.
+    ///
+    pub def intercalate[t :# Type -> Type : Foldable, a : Monoid : Eq](t: t[a], a: a): a =
+        Foldable.foldLeft((acc, x) ->
+        if (acc == Monoid.empty())
+            Monoid.combine(acc, x)
+        else
+            Monoid.combine(acc, Monoid.combine(a, x)), Monoid.empty(), t)
+
+    ///
+    /// Returns true if and only if `t` is empty.
+    ///
+    pub def isEmpty[t :# Type -> Type : Foldable, a](t: t[a]): Bool =
+        Foldable.foldLeft((_, _) -> false, true, t)
+
+    ///
+    /// Optionally finds the smallest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def minimum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> match acc {
+            case None => Some(x)
+            case Some(y) => Some(Order.min(x, y))
+        }, None, t)
+
+    ///
+    /// Optionally finds the smallest element of `t` according to the given comparator `cmp`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def minimumBy[t :# Type -> Type : Foldable, a](cmp: (a, a) -> Comparison, t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> match acc {
+            case None => Some(x)
+            case Some(y) => Some(Order.minBy(cmp, x, y))
+        }, None, t)
+
+    ///
+    /// Optionally finds the largest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def maximum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> match acc {
```

Comment:
There should be some useful combinators. Maybe Option.withDefault and Option.map? or Option.flatMap or something_

Reply:
I looked but it seems they all return None if one of the arguments are None. I'll check again later, there might be some combination of two functions.